### ISSUE TITLE
General API Questions

### ISSUE THREAD

**Issue by ngortheone** [slashdotted/libmelda#3]
Hey @slashdotted, thanks for the awesome library! I found it after reading your paper (which went a bit over my head :P ) and searched if there are any implementations. I consider myself lucky to find this in exactly the language I was interested in! 

After playing a little with libmelda I have a bunch of questions related to ergonomics and general API design:

1. To create a Melda instance one needs to pass in `Arc<RwLock<Adapter>>` which makes sense in multi-threaded or async scenarios, but presents a noticeable performance penalty for a single threaded use cases. I wonder whether Arc + RwLock is dictated by some internal melda needs or is it a way to avoid lifetime related issues and/or generics?

Looking at melda struct I can only guess that it was designed for use in heavy concurrency situations and allow interior mutability. Having each struct member under a separate RwLock/Mutex suggests that each field can be modified separately and concurrently without violating internal consistency. Is really possible to mutate `blocks` while someone else is mutating `documents` and keep melda consistent ?
```rust
pub struct Melda {
    documents: RwLock<BTreeMap<String, RwLock<RevisionTree>>>,
    data: RwLock<DataStorage>,
    stage: RwLock<Vec<Change>>,
    blocks: RwLock<BTreeMap<String, RwLock<Block>>>,
    array_descriptors_cache: Mutex<LruCache<Revision, ArrayDescriptor>>,
}

```

Would you be opposed to a redesign of the API to allow zero-cost single-threaded usecase? 

2. Each meld operation is followed by refresh, which raises a question why not combine these 2 into a single call? Are there scenarios where one could be useful without the other?
```rust
    melda_alice.meld(&melda_bob).expect("Failed to meld");
    melda_alice.refresh().expect("failed to refresh");

```

3. Refresh vs reload - these 2 look very similar on the surface. When each of these is supposed to be used?

4. Suppose I have a document that has high level of nesting:
```json
{
  "a": {
    "b": {
      "c": {
        "d": {"stuff": "foo"}
      }
    }
  }
}
```
How can I only update `d` without a need to serialize the entire document?
Current readme example shows this:
```rust
    // meanwhile Alice continues to make changes...
    let v = json!({ "software" : "MeldaDo", "version" : "1.0.0", "items♭" :
        [
            {"_id" : "alice_todo_02", "title" : "Take picture of our dog", "description" : "It must be a nice one"}
        ]
        })
     .as_object()
     .expect("Not an object")
     .clone();

    melda_alice.update(v).expect("Failed to update");

```
We serialize the entire document to `v` and then call `update(v)` method. Serializing big documents repetitively can be costly.
Is there a way to obtain a handle of `d` and update it separately ?

**Comment by slashdotted** [comment]
Hello @ngortheone , 
thank you for your comment! Concerning your questions:
1. There are situations where an application needs to keep a shared reference to the adapter in order to perform some operations that are not exposed through melda (for example, to check the status of a connection to a server, or to perform garbage collection).
2. Refresh can be expensive, especially if you have large deltas, because you need to load all the new ones and update the revision trees accordingly. If you are melding from different sources it is more efficient to refresh only afterwards. Moreover, there are situations where you just want to periodically meld the data from different sources but you don't want to update the local state. Meld is just a low level operation that syncs the data between adapters without updating the state. 
4. Reloads throws away the current state and reloads all delta blocks from the adapter, whereas refresh only loads new delta blocks.
5. You can update single objects (by _id) using **update_object(&self, uuid: &str, obj: Map<String, Value>) -> Result<()>** (this is the same method used by update when it detects that an object has been modified). However, only objects that are "extracted" by the flattening procedures can be managed that way. In your case you would need to assign an *_id* to the the object and ensure that the hierarchy is flattened (by adding ♭ to the name of the keys a, b, c, d):
```
{
  "a♭": {
    "b♭": {
      "c♭": {
        "d♭": {"_id": "myobject", "stuff": "foo"}
      }
    }
  }
}
```

cheers

A.

**Comment by ngortheone** [comment]
Is there a way to treat every nested object as marked with `♭` such that libmelda flattens everything?

**Comment by slashdotted** [comment]
Yes, but as of version 0.2.3 the only way is to change **is_flattened_field** (in _utils.rs_) and implement your own logic to determine whether a field is (to be) flattened or not. The current logic is actually very simple:
```rust
pub fn is_flattened_field(key: &str) -> bool {
    key.ends_with(FLATTEN_SUFFIX)
}
```
It could be possible to introduce some sort of "configuration" mechanism in the API to alter some of the default logic, but all such parameters would need to be (correctly) set each time you access the data or they would need to be stored as part of the state (perhaps in the first delta block?). 

Another part of the logic that could be made "configurable" is related to arrays: currently all updates to flattened arrays are stored as diffs (apart from the first revision). This solution minimizes the storage overhead when you add or remove some elements, but requires the client to have all delta blocks and packs in order to reconstruct the actual content of the array (hence if some data blocks or deltas are corrupted the whole data structure could be made unreadable). In previous versions of Melda you could choose to either store the full content or generate a diff (by prefixing the field name with Δ), but I've removed that option because I did not like the code I've written to handle that. Furthermore, apart from the situation where only the order of  the array changes, the client would still need all previous data blocks in order to determine the content of the object in the array.  I might reintroduce something in the future, but taking a different approach, for example, by adding an option to generate a full synchronization snapshot (no diffs) upon update.
